### PR TITLE
[Speculative decoding] Add periodic log with time spent in proposal/scoring/verification

### PR DESCRIPTION
**Pull request by cadedaniel**
* Log values whenever rejection sampler produces metrics
* Measures average time spent proposing each token, time spent scoring, and time spent verifiying.

### CODE REVIEW

File: vllm/spec_decode/spec_decode_worker.py [hunk 1]
```diff
@@ -912,3 +932,15 @@ def split_num_cache_blocks_evenly(scorer_cache_block_size_bytes: int,
...
+class Timer:
+
+    def __enter__(self):
+        self.start_time = time.time()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.end_time = time.time()
+        self.elapsed_time_s = self.end_time - self.start_time
+        self.elapsed_time_ms = self.elapsed_time_s * 1000
```
Comment (comaniac) [inline vllm/spec_decode/spec_decode_worker.py:946]: Put it in a more common space like utils?

File: vllm/spec_decode/spec_decode_worker.py [hunk 2]
```diff
@@ -725,6 +736,15 @@ def _create_output_sampler_list(
...
                 0].spec_decode_worker_metrics = maybe_rejsample_metrics
+
+            (average_time_per_proposal_tok_ms, scoring_time_ms,
+             verification_time_ms) = stage_times
+            logger.info(
+                "SpecDecodeWorker stage times: "
+                "average_time_per_proposal_tok_ms=%.02f "
+                "scoring_time_ms=%.02f verification_time_ms=%.02f",
+                average_time_per_proposal_tok_ms, scoring_time_ms,
+                verification_time_ms)
```
Comment (comaniac) [inline vllm/spec_decode/spec_decode_worker.py:747]: It looks like we will print this log every step? Did I miss something?

Comment (cadedaniel) [reply vllm/spec_decode/spec_decode_worker.py:747]: only when rejection sampler emits metrics, which is every 5s

File: vllm/spec_decode/spec_decode_worker.py (hunk 1 above)
Comment (cadedaniel) [reply vllm/spec_decode/spec_decode_worker.py:946]: ack

File: vllm/spec_decode/spec_decode_worker.py (hunk 2 above)
Comment (cadedaniel) [reply vllm/spec_decode/spec_decode_worker.py:746]: I'll add comment

**Review by comaniac — APPROVED** [review]
LGTM

**Review by rkooo567 — APPROVED** [review]
QQ; should we enable it only `disable_log_stats=False?`

**Comment by cadedaniel** [comment]
> QQ; should we enable it only disable_log_stats=False?

What is the motivation ?

**Comment by rkooo567** [comment]
> What is the motivation ?

oh because normally many users don't want to emit this kind of logs (and I thought that's why we have the disable_log_stats as an argument)

**Comment by cadedaniel** [comment]
Hmm good point. Let me see if I can pipe down that argument.

**Pull request merged**